fix MyPy type errors in api_fastapi/core_api/routes/public/* for SQLAlchemy 2 compatibility

rich7420 opened the pull request:
<!--
 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements.  See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership.  The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License.  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an
 "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 KIND, either express or implied.  See the License for the
 specific language governing permissions and limitations
 under the License.
 -->
related: #56735 

this PR is a part of this issue.

fix mypy type errors in public folder.

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->



<!-- Please keep an empty line above the dashes. -->
---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [airflow-core/newsfragments](https://github.com/apache/airflow/tree/main/airflow-core/newsfragments).

rich7420 committed:
fix mypy type errors in public

rich7420 committed:
remove comment

jason810496 reviewed (commented):
Thanks for the PR! However, the tests still need fix, thanks!

rich7420 committed:
fix  ruff error

rich7420 committed:
fix ci test

rich7420 committed:
fix ci test error

vincbeck commented on airflow-core/src/airflow/api_fastapi/core_api/routes/public/backfills.py:
```diff
@@ -214,7 +215,7 @@ def cancel_backfill(backfill_id: NonNegativeInt, session: SessionDep) -> Backfil
     # this is in separate transaction just to avoid potential conflicts
     session.refresh(b)
-    b.completed_at = timezone.utcnow()
+    b.completed_at = cast("UtcDateTime", timezone.utcnow())
```
This wont be needed when https://github.com/apache/airflow/pull/56780 is merged

vincbeck replied:
This is now merged, please rebase

vincbeck reviewed (commented):
Looks good overall! One comment and it needs to be rebased anyway

rich7420 committed:
Remove unnecessary UtcDateTime cast after PR #56780 merge

- Remove cast('UtcDateTime', timezone.utcnow()) as it's no longer needed
- Remove unused cast import
- This addresses reviewer feedback from vincbeck

uranusjr commented on airflow-core/src/airflow/api_fastapi/core_api/routes/public/assets.py:
```diff
@@ -244,7 +244,7 @@ def get_asset_aliases(
 
     return AssetAliasCollectionResponse(
-        asset_aliases=session.scalars(asset_aliases_select),
+        asset_aliases=list(session.scalars(asset_aliases_select)),
```
Instead of having to copy here, would it be possible to just change the model declaration from `list` to `Iterable` instead?

rich7420 replied:
yes! you're right

uranusjr commented on airflow-core/src/airflow/api_fastapi/core_api/routes/public/assets.py:
```diff
@@ -580,7 +580,7 @@ def get_dag_asset_queued_event(
 
     return QueuedEventResponse(
-        created_at=adrq.created_at,
+        created_at=cast("datetime", adrq.created_at),
```
Why is this needed?

rich7420 replied:
its not needed now.

uranusjr commented on airflow-core/src/airflow/api_fastapi/core_api/routes/public/assets.py:
```diff
@@ -535,7 +535,7 @@ def get_dag_asset_queued_events(
 
     dag_asset_queued_events_select, total_entries = paginated_select(statement=query)
-    adrqs = session.scalars(dag_asset_queued_events_select).all()
+    adrqs = list(session.scalars(dag_asset_queued_events_select))
```
Why does `.all()` not work?

rich7420 replied:
my mistake

rich7420 committed:
update

rich7420 committed:
remove unnecessary change

rich7420 committed:
ruff error

uranusjr commented on airflow-core/src/airflow/api_fastapi/core_api/routes/public/backfills.py:
I think we can change BackfillCollectionResponse too?

rich7420 replied:
no problem!I've tried.

rich7420 committed:
change BackfillCollectionResponse

jason810496 commented on airflow-core/src/airflow/api_fastapi/core_api/routes/public/dag_run.py:
```diff
@@ -626,7 +646,7 @@ def get_list_dag_runs_batch(
     )
 
-    dag_runs = session.scalars(dag_runs_select)
+    dag_runs = list(session.scalars(dag_runs_select))
```
I am wondering if there is a better way to resolve MyPy error without coping the result?
Since we add a lot of `list(...)`  in this PR.

jason810496 commented on airflow-core/src/airflow/api_fastapi/core_api/routes/public/dag_run.py:
```diff
@@ -205,14 +206,26 @@ def patch_dag_run(
+                raise HTTPException(
+                    status.HTTP_404_NOT_FOUND,
+                    f"DagRun with id: `{dag_run_pk}` was not found",
+                )
```
It seems we don't need to get `dag_run` again. Since we already select `dag_run` in line 167 and the not existed case at the beginning of the route.

```suggestion
```

jason810496 commented on airflow-core/src/airflow/api_fastapi/core_api/routes/public/dag_run.py:
```diff
@@ -205,14 +206,26 @@ def patch_dag_run(
+        raise HTTPException(
+            status.HTTP_404_NOT_FOUND,
+            f"DagRun with id: `{dag_run_pk_final}` was not found",
+        )
```
Same as here, maybe `session.refersh(dag_run)`  _might_ be enough?

jason810496 commented on airflow-core/src/airflow/api_fastapi/core_api/routes/public/xcom.py:
```diff
@@ -193,8 +194,10 @@ def get_xcom_entries(
+    xcoms = list(session.scalars(query))
+    return XComCollectionResponse(
+        xcom_entries=[XComResponse.from_orm(xcom) for xcom in xcoms], total_entries=total_entries
+    )
```
For this case, we even copy the result twice.
Not sure if it's possible just to use `cast` to resolve the MyPy issue?

jason810496 reviewed (commented):
Nice! Thanks for the PR!

rich7420 committed:
to modify extra codes

jason810496 commented on airflow-core/src/airflow/api_fastapi/core_api/routes/public/dag_versions.py:
```diff
@@ -122,7 +122,7 @@ def get_dag_versions(
         session=session,
     )
-    dag_versions = session.scalars(dag_versions_select)
+    dag_versions = list(session.scalars(dag_versions_select))
```
There are still quite a few place that use `list(...)`for copying result.
It would be nice to check all similar usage in the PR and try whether could we avoid coping.
Thanks!

rich7420 replied:
@jason810496 thanks for the feedback! 
I've tried to remove all similar situation using list() and cast().

jason810496 commented on airflow-core/src/airflow/api_fastapi/core_api/routes/public/hitl.py:
```diff
@@ -286,7 +286,10 @@ def get_hitl_details(
+    hitl_details_raw = list(session.scalars(hitl_detail_select))
+
+    # Convert SQLAlchemy models to Pydantic models
+    hitl_details = [HITLDetail.model_validate(hitl_detail) for hitl_detail in hitl_details_raw]
```
Is it required to convert to Pydantic models to solve the MyPy error here?

jason810496 reviewed (commented):
Thanks for the update!
Please don't hesitate to resolve the comments that you already addressed.

rich7420 committed:
remove list() and cast

rich7420 committed:
remove extra list() and cast

rich7420 committed:
remove cast

rich7420 committed:
fix errors

rich7420 committed:
Remove exclude_original parameter from partial_subset call

rich7420 committed:
Remove exclude_original parameter from partial_subset call

rich7420 committed:
remove extra modification

rich7420 committed:
Merge branch 'main' into AIRFLOW56735-public

rich7420 committed:
Fix end-of-file-fixer: Add missing newlines to public API route files

jason810496 reviewed (commented):
It seems like a bad rebase? Since most of the changes are missing after push.

jason810496 commented on airflow-core/src/airflow/api_fastapi/core_api/routes/public/xcom.py:
```diff
@@ -193,8 +194,10 @@ def get_xcom_entries(
-    return XComCollectionResponse(xcom_entries=xcoms, total_entries=total_entries)
+    xcoms = session.scalars(query).all()
+    return XComCollectionResponse(
+        xcom_entries=[XComResponse.from_orm(xcom) for xcom in xcoms], total_entries=total_entries
```
Will MyPy raise error if we just use `session.scalars(query)` ?
```suggestion
        xcom_entries=session.scalars(query), total_entries=total_entries
```

rich7420 committed:
Update xcom.py

Co-authored-by: LIU ZHE YOU <68415893+jason810496@users.noreply.github.com>

rich7420 committed:
fix ruff error
